packages/chrome-offscreen-worker/src/worker.ts

```diff
@@ -0,0 +1,145 @@
+
+  constructor(...[scriptURL, options]: ConstructorParameters<typeof Worker>) {
+    this.params = [
+      String(scriptURL),
+      {
+        name: `${this.constructor.name} ${Date.now()} ${String(scriptURL)}`,
+        ...options,
+      },
+    ];
+
+    if (!OffscreenWorker.control) {
+      throw new Error(
+        `${this.constructor.name + '.configure'} must be called before constructing ${this.constructor.name}`,
+      );
+    }
+
+    this.workerPort = OffscreenWorker.control.constructWorker(...this.params);
+
+    const { port1, port2 } = new MessageChannel();
+    this.internalTarget = port1;
+    this.externalTarget = port2;
+
+    this.externalTarget.addEventListener(
+      'error',
+      evt => void this.onerror?.call(this, evt as ErrorEvent),
+    );
+    this.externalTarget.addEventListener('message', evt => void this.onmessage?.call(this, evt));
+    this.externalTarget.addEventListener(
+      'messageerror',
+      evt => void this.onmessageerror?.call(this, evt),
+    );
+
+    void this.workerPort.then(
+      port => {
+        console.log(this.params[1].name, 'got chromePort');
+        port.onMessage.addListener(this.workerListener);
+
+        this.internalTarget.addEventListener('error', this.parentListener);
+        this.internalTarget.addEventListener('message', this.parentListener);
+        this.internalTarget.addEventListener('messageerror', this.parentListener);
+
+        this.internalTarget.start();
+      },
+      (error: unknown) => this.internalTarget.dispatchEvent(new ErrorEvent('error', { error })),
+    );
+  }
+
+  private workerListener = (json: unknown, port: chrome.runtime.Port) => {
+    console.debug(this.params[1].name, 'workerListener', json, port.name);
+    if (isOffscreenWorkerEventMessage(json) && isOffscreenWorkerEvent(json.init, json.type)) {
+      switch (json.type) {
+        case 'error':
+          this.internalTarget.dispatchEvent(new ErrorEvent(json.type, json.init));
+          return;
+        case 'message':
+        case 'messageerror':
+          this.internalTarget.dispatchEvent(new MessageEvent(json.type, json.init));
+          return;
+        default:
+          //console.warn('Dispatching unknown event', json);
+          //this.internalTarget.dispatchEvent(new Event(json.type, json.init));
+          throw new Error('Unknown event in worker output', {
+            cause: { type: Boolean(json) && json.type, message: json, port: port.name },
+          });
+      }
+    }
+  };
+
+  private parentListener = (evt: Event) => {
+    console.debug(this.params[1].name, 'parentListener', [evt]);
+    const unknownMessage = { type: evt.type, init: { ...evt } };
+    switch (evt.type) {
+      case 'error':
+      case 'message':
+      case 'messageerror':
+        void this.workerPort.then(port => port.postMessage(unknownMessage as OffscreenWorkerEvent));
+        return;
+      default:
+        throw new Error('Unknown event from parent', { cause: evt });
+    }
+  };
```

Comment:
this accepts pretty much anything. it probably needs some validation, but again, technically callers could dispatch pretty much anything.